src/main/java/com/commercetools/sync/inventories/InventorySync.java

```diff
@@ -217,32 +254,57 @@ public InventorySyncStatistics getStatistics() {
                                                                              @Nonnull final Map<String, String>
                                                                                  supplyChannelKeyToId) {
         if (syncOptions.shouldEnsureChannels()) {
-            final List<String> missingChannelsKeys = drafts.stream()
-                .map(ChannelKeyExtractor::extractChannelKey)
-                .distinct()
-                .filter(Objects::nonNull)
-                .filter(key -> !supplyChannelKeyToId.containsKey(key))
-                .collect(toList());
-            final List<CompletableFuture<Channel>> creationStages = missingChannelsKeys.stream()
-                .map(this::createMissingSupplyChannel)
-                .map(newChannelStage -> newChannelStage
-                    .handle((newChannel, exception) -> {
-                        if (newChannel != null) {
-                            supplyChannelKeyToId.put(newChannel.getKey(), newChannel.getId());
-                        }
-                        return newChannel;
-                    }))
-                .map(CompletionStage::toCompletableFuture)
+            final List<String> missingChannelsKeys = findMissingChannelsKeys(drafts, supplyChannelKeyToId);
+            final List<CompletableFuture<Void>> newChannelsFutures = createSupplyChannelsOfKeys(missingChannelsKeys)
+                .stream()
+                .map(newChannelFuture -> newChannelFuture
+                    .thenAccept(newChannelOptional -> newChannelOptional.ifPresent(
+                        newChannel -> supplyChannelKeyToId.put(newChannel.getKey(), newChannel.getId()))))
```

Comment:
looks like this modifies the map as a side effect in every future. Despite being hard to read,
potentially it could also cause concurrency problems if the code would be parallel.
I'd suggest to modify map AFTER all channels are created and ready in a single shot.